kube-runtime/src/lease.rs

```diff
@@ -0,0 +1,287 @@
+//! Acquire and hold a lease (advisory lock)
+
+use k8s_openapi::{
+    api::coordination::v1::{Lease as KubeLease, LeaseSpec as KubeLeaseSpec},
+    apimachinery::pkg::apis::meta::v1::MicroTime,
+    chrono::{Local, Utc}
+};
+
+use kube_client::api::{ ObjectMeta, Patch, PatchParams, PostParams };
+
+use std::time::Duration;
+use tokio::{sync::oneshot::Sender, task::JoinHandle};
+use kube_client::Api;
+use kube_client::error::ErrorResponse;
+
+const LEASE_DURATION_SECONDS: u64 = 5;
+
+/// This implementation provides a guard for routinely renewing an Lease object until dropped. Once
+/// dropped the spawned renewal task is allowed to complete. This implementation also allows callers
+/// to poll the availability of a Lease, and also perform a blocking call to acquire a lease.
+/// Acquiring an existing Lease is done with a compare and swap, relying on the Kubernetes API to
+/// ensure excusive ownership of Lease.
+///
+/// Leases are "acquired", then periodically "renewed", and they track the "owner" of the lease. The renewal
+/// frequency is not configurable but leases are released as soon as the guard is dropped.
+/// When a lease is no longer needed, all the properties are nullified so other processes can take over.
+///
+/// A Lease is a great tool for signaling a process is underway -- this ensures cooperating software should
+/// wait before continuing. Leases are named and namespaced objects.
+///
+/// Future work should attach Events to a Lease so `kubernetes describe` can make a useful history of the
+/// lease.
+///
+/// How to use a lease:
+///
+///
+///
+pub struct Lease {
+    join_handle: JoinHandle<()>,
+    sender: Sender<()>,
+}
+
+impl Lease {
+    pub fn lease_name() -> &'static str {
+        "cousteau"
+    }
+
+    pub fn lease_duration() -> u64 {
+        LEASE_DURATION_SECONDS
+    }
+
+    pub async fn available(
+        kube_api_client: kube_client::Client,
+        ns: &str,
+        lease_name: &str,
+    ) -> Result<bool, String> {
+        let lease_client: Api<KubeLease> =
+            kube_client::api::Api::namespaced(kube_api_client.clone(), ns);
+
+        let get_lease = lease_client.get(lease_name).await;
+
+        match get_lease {
+            Err(kube_client::Error::Api(ErrorResponse { code: 404, .. })) => Ok(true),
+            Err(err) => Err(format!("{:#?}", err)),
+            Ok(lease) => Ok(Self::lease_expired(&lease)),
+        }
+    }
+
+    pub async fn acquire_or_create(
```

Comment:
I do feel that these big functions here should be split up.

`acquire` and `create` would be fine on their own given some state and it would make the logic of the looping better (if that's how we want to handle it)
[leaderelection.go](https://github.com/kubernetes/client-go/blob/master/tools/leaderelection/leaderelection.go) has `acquire`, `release` and `renew` for reference.